etl/src/v2/concurrency/stream.rs

```diff
@@ -1,60 +1,27 @@
+use crate::v2::concurrency::shutdown::{ShutdownResult, ShutdownRx};
+use crate::v2::config::batch::BatchConfig;
 use core::pin::Pin;
 use core::task::{Context, Poll};
 use futures::{ready, Future, Stream};
 use pin_project_lite::pin_project;
-use tokio::sync::watch;
-use tokio::time::{sleep, Sleep};
 use tracing::info;
 
-use crate::v2::config::batch::BatchConfig;
-
-/// A trait that determines whether an item in a stream marks the end of a batch.
-///
-/// Types implementing this trait can signal when they should be the last item in a batch,
-/// allowing for intelligent batch boundary decisions in streaming operations.
-pub trait BatchBoundary: Sized {
-    /// Returns `true` if this item should be the last in its batch, that is, the item is on
-    /// a boundary.
-    ///
-    /// This method is used by [`BoundedBatchStream`] to determine when to emit a batch.
-    fn is_on_boundary(&self) -> bool;
-}
-
-impl<T: BatchBoundary, E> BatchBoundary for Result<T, E> {
-    fn is_on_boundary(&self) -> bool {
-        match self {
-            Ok(v) => v.is_on_boundary(),
-            // We return true since in case of error we want to fail fast, since the batch is
-            // anyway going to fail.
-            Err(_) => true,
-        }
-    }
-}
-
 // Implementation adapted from:
 //  https://github.com/tokio-rs/tokio/blob/master/tokio-stream/src/stream_ext/chunks_timeout.rs
 pin_project! {
     /// A stream adapter that batches items based on size limits and timeouts.
     ///
     /// This stream collects items from the underlying stream into batches, emitting them when either:
-    /// - The batch reaches its maximum size and the last item indicates it's a batch boundary
-    /// - A timeout occurs and the last item indicates it's a batch boundary
-    /// - The stream is forcefully stopped
-    ///
-    /// The stream guarantees that batches will end on items that return `true` from
-    /// [`BatchBoundary::is_on_boundary`], unless the stream is forcefully stopped.
-    ///
-    /// # Implementation Details
-    ///
-    /// The implementation is adapted from Tokio's chunks_timeout stream extension.
+    /// - The batch reaches its maximum size
+    /// - A timeout occurs
     #[must_use = "streams do nothing unless polled"]
     #[derive(Debug)]
-    pub struct BoundedBatchStream<B: BatchBoundary, S: Stream<Item = B>> {
+    pub struct BatchStream<B, S: Stream<Item = B>> {
```

Comment:
Removed the `BatchBoundary` as agreed with @imor, since we can't buffer data indefinitely in-memory and Postgres doesn't give us any upper bounds in transaction size.